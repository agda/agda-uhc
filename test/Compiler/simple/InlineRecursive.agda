{-# OPTIONS -v treeless.opt:20 #-}
-- Test that inlining a recursive function doesn't throw
-- the compiler into a loop.
module _ where

open import Common.Prelude

f : Nat â†’ Nat
f zero = zero
f (suc n) = f n
{-# INLINE f #-}

main : IO Unit
main = printNat (f 4)
