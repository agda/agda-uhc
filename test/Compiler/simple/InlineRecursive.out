EXECUTED_PROGRAM

ret > ExitSuccess
out > InlineRecursive.f =
out >   λ a →
out >     case a of
out >       0 → 0
out >       _ → let b = a - 1 in InlineRecursive.f b
out > InlineRecursive.main =
out >   Common.IO.printNat
out >     (let a = 4 in
out >      case a of
out >        0 → 0
out >        _ → InlineRecursive.f 3)
out > 0
